File: src/code/my_josie_data_prep/create_raw_tokens.py
import os
import sys
import pyspark

from code.utils.settings import DefaultPath

def create_raw_tokens(input_set_file, output_raw_tokens_file, single_txt=True, spark_context=None):
    if not spark_context:
        conf = pyspark.SparkConf() \
            .setAppName('CreateIndex') \
                # .set('spark.executor.memory', '100g') \
                # .set('spark.driver.memory', '5g')
        spark_context = pyspark.SparkContext(conf=conf)
    
    skip_tokens = set()

    sets = spark_context.textFile(input_set_file) \
        .map(
            lambda line: line.split(',')
        ) \
        .map(
            lambda line: (
                int(line[0]),
                [token for token in line[1:] if token not in skip_tokens]
        )
    )

    if single_txt:
        sets = sets \
            .flatMap(
                lambda sid_tokens: \
                    [
                        (token, sid_tokens[0]) 
                        for token in sid_tokens[1]
                    ]
                ) \
                    .map(
                        lambda token_sid: f'{token_sid[0]} {token_sid[1]}\n'
                    ).collect()
        with open(output_raw_tokens_file, 'w') as f:
            f.writelines(sets)
    else:
        sets \
            .flatMap(
                lambda sid_tokens: 
                    [
                        (token, sid_tokens[0]) 
                        for token in sid_tokens[1]
                    ]
                ) \
                    .map(
                        lambda token_sid: f'{token_sid[0]} {token_sid[1]}'
                    ).saveAsTextFile(output_raw_tokens_file)


if __name__ == '__main__':
    ROOT_TEST_DIR = DefaultPath.data_path.wikitables + 'threshold_r5-c2-a50/josie-test'
    input_set_file = ROOT_TEST_DIR + 'sloth_tables_n10.set'
    output_raw_tokens_file = ROOT_TEST_DIR + 'sloth_tables_n10.raw-token'

File: josie-indexing-scala/src/main/scala/CreateRawTokens.scala
import org.apache.hadoop.conf.Configuration
import org.apache.spark.SparkConf
import org.apache.spark.SparkContext
import org.apache.log4j.{Level, Logger}


object CreateRawTokens {
  def main(args: Array[String]): Unit = {
    Logger.getLogger("org").setLevel(Level.ERROR)
    val conf = new SparkConf().setAppName("CreateIndex").setMaster("local[1]")
    // .set("spark.executor.memory", "100g")
    // .set("spark.driver.memory", "5g")
    val sc = new SparkContext(conf)

    print("Initial arguments: ")
    print(args.mkString(", "))
    print("\n")

    // necessary to run the jar file
    val hadoopConfig: Configuration = sc.hadoopConfiguration
    hadoopConfig.set("fs.hdfs.impl", classOf[org.apache.hadoop.hdfs.DistributedFileSystem].getName)
    hadoopConfig.set("fs.file.impl", classOf[org.apache.hadoop.fs.LocalFileSystem].getName)

    // The input sets, preprocessed to remove numerical values
    val setFile = args(0)

    // The output token file.
    val rawTokensFile = args(1)

    // The tokens to skip.
    val skipTokens = Set("##################################")

    val sets = sc.textFile(setFile).map(line => line.split('|'))
      .map(line => (line(0).toInt, line.drop(1).filter(
        token => !skipTokens(token))))
    sets.flatMap { case (sid, tokens) =>
      tokens.map(token => (token, sid))
    }.map{
      case (token, sid) => s"$token $sid"
    }.saveAsTextFile(rawTokensFile)
  }
}
